### PR TITLE
Fix double callback injection issue with LangChain streaming

### PR DESCRIPTION
**Pull request by dbczumar**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/dbczumar/mlflow/pull/12037?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12037/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12037
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Fix double callback injection issue with LangChain streaming

### How is this PR tested?

- [ ] Existing unit/integration tests
- [X] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [X] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

File: tests/langchain/test_langchain_autolog.py [hunk 1]
```diff
@@ -891,5 +892,9 @@ def test_langchain_autolog_callback_injection_in_stream(invoke_arg, generate_con
     expected_logs = ["chain_start", "chain_end"]
     if isinstance(callbacks, BaseCallbackManager):
         assert callbacks.handlers[0].logs == expected_logs
+        assert (
+            sum(isinstance(handler, MlflowLangchainTracer) for handler in callbacks.handlers) == 1
+        )
```
Comment (dbczumar) [inline tests/langchain/test_langchain_autolog.py:897]: This was previously `2`, which was an error.

Technically, we shouldn't be mutating the list of callbacks that the user provides to the `stream()` call. In the interest of time, I'm adjusting this test for now and will file a ticket for fixing the erroneous modification of the user's callback list and adjust the test case

File: mlflow/langchain/_langchain_autolog.py [hunk 2]
```diff
@@ -257,13 +257,30 @@ def _inject_callbacks(original_callbacks, new_callbacks):
...
+            for original_callback in original_callbacks
+        )
+
     if isinstance(original_callbacks, BaseCallbackManager):
         for callback in new_callbacks:
-            original_callbacks.add_handler(callback)
+            if not _is_new_callback_already_in_original_callbacks(
+                callback, original_callbacks.handlers
+            ):
+                original_callbacks.add_handler(callback)
```
Comment (dbczumar) [inline mlflow/langchain/_langchain_autolog.py:271]: We shouldn't be modifying `original_callbacks` either. Filed a follow-up ticket to fix that.

**Review by serena-ruan — APPROVED** [review]
LGTM! Thanks for the fix!

**Comment by dbczumar** [comment]
@mlflow-automation autoformat

**Comment by dbczumar** [comment]
Failures are unrelated. Merging.

**Pull request merged**